src/main/scala/strawman/collection/mutable/ArrayDeque.scala

```diff
@@ -0,0 +1,349 @@
+  *
+  *  @define Coll `mutable.ArrayDeque`
+  *  @define coll array deque
+  *  @define thatinfo the class of the returned collection. In the standard library configuration,
+  *    `That` is always `ArrayDeque[B]` because an implicit of type `CanBuildFrom[ArrayDeque, B, ArrayDeque[B]]`
+  *    is defined in object `ArrayDeque`.
+  *  @define bfinfo an implicit value of class `CanBuildFrom` which determines the
+  *    result class `That` from the current representation type `Repr`
+  *    and the new element type `B`. This is usually the `canBuildFrom` value
+  *    defined in object `ArrayDeque`.
+  *  @define orderDependent
+  *  @define orderDependentFold
+  *  @define mayNotTerminateInf
+  *  @define willNotTerminateInf
+  */
+@SerialVersionUID(1L)
+class ArrayDeque[A] private[ArrayDeque](
+    private[ArrayDeque] var array: Array[AnyRef],
+    private[ArrayDeque] var start: Int,
+    private[ArrayDeque] var end: Int
+) extends mutable.AbstractBuffer[A]
+    with mutable.Buffer[A]
+    with generic.GenericTraversableTemplate[A, ArrayDeque]
+    with mutable.BufferLike[A, ArrayDeque[A]]
+    with mutable.IndexedSeq[A]
+    with mutable.IndexedSeqOptimized[A, ArrayDeque[A]]
+    with mutable.Builder[A, ArrayDeque[A]]
+    with Serializable {
+
+  private[this] var mask = 0   // modulus using bitmask since array.length is always power of 2
+  set(array, start, end)
+
+  def this(initialSize: Int = ArrayDeque.DefaultInitialSize) = this(ArrayDeque.alloc(initialSize), 0, 0)
+
+  override def apply(idx: Int) = {
+    ArrayDeque.checkIndex(idx, this)
+    get(idx).asInstanceOf[A]
+  }
+
+  override def update(idx: Int, elem: A) = {
+    ArrayDeque.checkIndex(idx, this)
+    set(idx, elem.asInstanceOf[AnyRef])
+  }
+
+  override def +=(elem: A) = {
+    sizeHint(size)
+    appendAssumingCapacity(elem)
+  }
+
+  override def +=:(elem: A) = {
+    sizeHint(size)
+    prependAssumingCapacity(elem)
+  }
+
+  @inline private[ArrayDeque] def appendAssumingCapacity(elem: A): this.type = {
+    array(end) = elem.asInstanceOf[AnyRef]
+    end = end_-(-1)
+    this
+  }
+
+  @inline private[ArrayDeque] def prependAssumingCapacity(elem: A): this.type = {
+    start = start_+(-1)
+    array(start) = elem.asInstanceOf[AnyRef]
+    this
+  }
+
+  override def ++=:(elems: TraversableOnce[A]): this.type = {
+    // Note this is faster than super.++=: because this does not need to convert TraversableOnce to a Traversable
+    if (elems.isEmpty) return this
+    elems.sizeHintIfCheap match {
+      case srcLength if srcLength >= 0 && 2*(srcLength + this.length) >= mask =>
+        val finalLength = srcLength + this.length
+        val array2 = ArrayDeque.alloc(finalLength)
+        elems.copyToArray(array2.asInstanceOf[Array[A]])
+        copySliceToArray(srcStart = 0, dest = array2, destStart = srcLength, maxItems = size)
+        set(array = array2, start = 0, end = finalLength)
+      case _ =>
+        val copy = clone()
```

Comment:
But you know the endpoints.  You can handle it manually.  I guess this does reduce the chance of implementation error.